### ISSUE TITLE
Wrong mime type

### ISSUE THREAD

**Issue by alssdllc** [mscdex/busboy#362]
I'm still trying to understand how all of this works. The issue is I record a video with MediaRecorder in the browser and submit it to my node server which uses multer (which uses busboy). 

In chrome the content-type is: **video/mp4;codecs=avc1,opus** even when I specify **video/mp4**.
In safari the content-type is **video/mp4**.

busboy returns the proper mimetype for safari but returns **text/plain** for chrome. The safari made video is saved properly and plays. The chrome made video does not play properly. I'm assuming it's corrupt or something.

**Comment by mscdex** [comment]
I think that's an issue you'll have to dive into yourself because `busboy` can only give you what was given to it. It doesn't modify content types or anything like that. The most important thing to ensure is that you're giving `busboy` multipart/form-data data and not video data directly.

**Comment by alssdllc** [comment]
You are correct that the content-type is not modified. But isn't the content-type used to determine the mime type?

When I look through some of the code. _I am comparing the safari run and the chrome run._ 

They both hit this code _(in multipart.js)_:
```
if (header['content-type']) {
    const conType = parseContentType(header['content-type'][0]);
    if (conType) {
        partType = `${conType.type}/${conType.subtype}`;
        if (conType.params && typeof conType.params.charset === 'string')
            partCharset = conType.params.charset.toLowerCase();
    }
}
```

For safari `header['content-type'][0]` is equal to `video/mp4`
For chrome `header['content-type'][0]` is equal to `video/mp4;codecs=avc1,opus`

For safari `conType` is a proper object:
```
params = {}
subtype = 'mp4'
type = 'video'
```
For chrome `conType` is `undefined`.

I would imagine that is part of the problem.

**Comment by aydincandan** [comment]
Well, how nice, Chrome gives more, it is not missing or wrong.

**Comment by ChristianUlbrich** [comment]
This is a bug. It happens, because of the comma in the codecs _parameter value_. Having dived way too deep into RFCs and standards, it seems that _comma-separated_ **_values_** for **_parameters_** are allowed. They are certainly for `video/mp4` according to [RFC6381](https://www.rfc-editor.org/rfc/rfc6381.html#section-3.6).

For an easy repro head over to https://codepen.io/ChristianUlbrich/pen/zxvgKLz?editors=1111 (directly taken from _source_)

The problem is _presumably_ in `parseContentTypeParams`, AI says:

> It’s because the parser only allows unquoted “token” chars in parameter values. A comma , (ASCII 44) is not a token (TOKEN[44] === 0). The unquoted value stops at the comma (avc1), and the next char is not a semicolon ;, so the function treats it as malformed and returns undefined.

This looks plausible™ and hints at an incorrect, not rfc-compliant implementation of _parameter values_ for _media types_. As the various _"parsers"_ are hand-crafted loops, fixing this w/o affecting other stuff is not so trivial, well at least not for me. :)

I'd go either with a _real_ (A)BNF parser and supplying the [BNF rules for The Content-Type Header Field](https://www.w3.org/Protocols/rfc1341/4_Content-Type.html) **OR** use a way more trivial, "direct" parser, that is much easier to read, than the current loop and magic token trickery.

**Comment by mscdex** [comment]
> They are certainly for `video/mp4` according to [RFC6381](https://www.rfc-editor.org/rfc/rfc6381.html#section-3.6).

Notice how in the examples you linked to the values containing commas are only ever enclosed within quotes.

> the [BNF rules for The Content-Type Header Field](https://www.w3.org/Protocols/rfc1341/4_Content-Type.html)

If you look at these rules more closely you'll notice that commas are only allowed inside a quoted value.
